### PR TITLE
[Dependency Scanning] Make dependency scanner cache specific to a given target triple.

### PR DESCRIPTION
**Pull request by artemcm**
In order to do so, first refactor the dependency kinds to model the main modules of individual scanning actions as separate dependency kind: `SwiftSource`. These kinds of modules differ from `SwiftTextual` modules in that they do not have an interface and have source-files. It is cleaner to enforce this distinction with types, instead of checking for interface optionality everywhere, as we did prior to this change.

Then, make `GlobalModuleDependenciesCache` aware of the current scanning action's target triple and only resolve cached dependencies that came from scanning actions with the same target triple.

This change means that the `GlobalModuleDependenciesCache` must be configured with a specific target triple for every scanning action, and it will only resolve previously-found dependencies from previous scannig actions using the exact same triple **only**.

Furthermore, the `GlobalModuleDependenciesCache` now separately tracks source-file-based module dependencies as those represent main Swift modules of previous scanning actions, and we must be able to resolve those regardless of the target triple.

This change is largely required due to the introduction of the `-clang-target` optimization which means that the driver is no longer re-scanning Clang modules against different target sets in a given module's build graph. 

Resolves rdar://83105455

### CODE REVIEW

File: include/swift-c/DependencyScan/DependencyScan.h [hunk 1]
```diff
@@ -49,6 +49,8 @@ typedef struct {
 } swiftscan_string_set_t;
 
 typedef enum {
+  // This dependency info encodes two ModuleDependencyKind types:
```
Comment (artemcm) [inline include/swift-c/DependencyScan/DependencyScan.h:52]: For now, preserve the format used by libSwiftScan. Once we teach the clients about the new dependency kind, we can then add it here.

**Comment by artemcm** [comment]
Existing scanner tests ensure we do not regress on current functionality.
Cross-triple test will be added in the swift-driver repository.

**Comment by artemcm** [comment]
@swift-ci please test

**Comment by artemcm** [comment]
@swift-ci please test

File: include/swift/AST/ModuleDependencies.h [hunk 2]
```diff
@@ -37,7 +37,8 @@ class Identifier;
 /// Which kind of module dependencies we are looking for.
 enum class ModuleDependenciesKind : int8_t {
   FirstKind,
-  SwiftTextual = FirstKind,
+  SwiftInterface = FirstKind,
+  SwiftSource,
```
Comment (nkcsgexi) [inline include/swift/AST/ModuleDependencies.h:41]: Minor nitpick: new cases are generally preferred to be appended at the last one to preserve stability of the existing cases.

File: include/swift/AST/ModuleDependencies.h [hunk 3]
```diff
@@ -152,14 +153,51 @@ class SwiftTextualModuleDependenciesStorage :
...
+  Optional<std::string> bridgingHeaderFile;
+
+  /// Swift source files that are part of the Swift module, when known.
+  std::vector<std::string> sourceFiles;
+
+  /// Source files on which the bridging header depends.
+  std::vector<std::string> bridgingSourceFiles;
+
+  /// (Clang) modules on which the bridging header depends.
+  std::vector<std::string> bridgingModuleDependencies;
```
Comment (nkcsgexi) [inline include/swift/AST/ModuleDependencies.h:185]: Minor nitpick: can we refactor these data entries to a struct and share it with `SwiftTextualModuleDependenciesStorage`?

**Review by nkcsgexi — APPROVED** [review]
Having a separate dependency kind for source file seems reasonable to me. Only minor nitpicks.

**Comment by artemcm** [comment]
@swift-ci please test

**Pull request merged**